cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
+
+        elif isinstance(lhs, GlobalConstraint) and lhs.name not in supported:
+            raise ValueError("Linearization of `lhs` not supported, run `cpmpy.transformations.decompose_global.decompose_global() first")
+
+        if is_num(lhs) or (isinstance(lhs, Operator) and lhs.name in {"sum","wsum"}):
+            # bring all vars to lhs
+            if isinstance(rhs, _NumVarImpl):
+                if isinstance(lhs, Operator) and lhs.name == "sum":
+                    lhs, rhs = sum([1 * a for a in lhs.args]+[-1 * rhs]), 0
+                else:
+                    lhs, rhs = lhs + -1*rhs, 0
+
+            # bring all const to rhs
+            if lhs.name == "sum":
+                new_args = []
+                for i, arg in enumerate(lhs.args):
+                    if is_num(arg):
+                        rhs -= arg
+                    else:
+                        new_args.append(arg)
+                lhs = Operator("sum", new_args)
+
+            elif lhs.name == "wsum":
+                new_weights, new_args = [],[]
+                for i, (w, arg) in enumerate(zip(*lhs.args)):
+                    if is_num(arg):
+                        rhs -= w * arg
+                    else:
+                        new_weights.append(w)
+                        new_args.append(arg)
+                lhs = Operator("wsum",[new_weights, new_args])
+
+
+        if isinstance(lhs, Operator) and lhs.name == "mul" and is_num(lhs.args[0]):
+            # convert to wsum
+            lhs = Operator("wsum",[[lhs.args[0]],[lhs.args[1]]])
+
+        # now fix the comparisons themselves
+        if cpm_expr.name == "<":
+            new_rhs, cons = get_or_make_var(rhs - 1) # if rhs is constant, will return new constant
+            return [lhs <= new_rhs] + cons
+        if cpm_expr.name == ">":
+            new_rhs, cons = get_or_make_var(rhs + 1) # if rhs is constant, will return new constant
+            return [lhs >= new_rhs] + cons
+        if cpm_expr.name == "!=":
+            # Special case: BV != BV
+            if isinstance(lhs, _BoolVarImpl) and isinstance(rhs, _BoolVarImpl):
+                return [lhs + rhs == 1]
+
+            if reified or (isinstance(lhs, (Operator, GlobalConstraint)) and lhs.name not in {"sum","wsum"}):
+                # lhs is sum/wsum and rhs is contant OR
+                # lhs is GenExpr and rhs is constant or var
+                #  ... what requires less new variables?
+                # Big M implementation
+                z = boolvar()
+                # Calculate bounds of M = |lhs - rhs| + 1
+                _, M = (abs(lhs - rhs) + 1).get_bounds()
+
+                cons = [lhs - M*z <= rhs-1, lhs - M*z >= rhs-M+1]
+                return linearize_constraint(flatten_constraint(cons), supported=supported, reified=reified)
```

Comment:
Why `flatten_constraint`?

Reply:
In theory, rhs can be a variable at this point if lhs is not a sum/wsum.
So for example when linearizing `max(x,y,z) != p` we cannot bring `p` to the lhs...
Notice however that in practice this does never occur as we rewrite `max(x,y,z) != p` as `[max(x,y,z) == iv, p != iv]` using `only_numexpr_equality`. But it is not a precondition for the `linearize` function...